### PR TITLE
firewall: Show any included services on a service

### PR DESCRIPTION
**Pull request by Gundersanne**
~~blocked on:~~
~~- [ ] https://bugzilla.redhat.com/show_bug.cgi?id=1737045~~

using `org.fedoraproject.FirewallD1.config.service.getSettings2()`

---

Fixes rhbz#1721548.

In the image below, the DNS service doesn't have any included services, freeipa4 does.

image

### CODE REVIEW

**Comment by marusak** [comment]
Cool stuff :)
I was at first just rather confused :D Have you talked about this with @garrett or @andreasn ?
I know that not always there has to be `Details` tab. But why the second one has `FreeIPA 4 server` (itself) but first has `Details` and not `DNS`? 
Maybe it would be better to have just one tab `Included Services` and having some listing in there?
What do you think @garrett ?

**Comment by Gundersanne** [comment]
@marusak yea I wasn't sure about the design at all. More consistency there would be good I agree.

**Comment by garrett** [comment]
This is really cool.

I've always thought it's a little odd to have 1 tab for details.

I wonder if it makes sense to remove the tabs altogether and show all the information below.

Is there a service that pulls in more services than FreeIPA currently?

**Comment by Gundersanne** [comment]
@garrett I think freeipa-4 includes quite a lot. Most are about 4-5.

**Comment by marusak** [comment]
Needs to rebase to master since #12367 changed tests names

File: pkg/networkmanager/firewall-client.js [hunk 1]
```diff
@@ -155,7 +155,18 @@ function getServices() {
...
+                .then(services => {
+                    let promises = [];
+                    for (let s of services) {
+                        firewall.enabledServices.add(s.id);
+                        if (s.includes.length)
+                            promises.push(fetchServiceInfos(s.includes));
+                    }
+                    // We can't use Promise.all() here until cockpit is able to dispatch es2015 promises
+                    // https://github.com/cockpit-project/cockpit/issues/10956
+                    // eslint-disable-next-line cockpit/no-cockpit-all
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: This should actually be obsolete now. But let's keep it here, and we can drop all cockpit.all() in a separate PR.

File: test/verify/check-networking-firewall [hunk 2]
```diff
@@ -181,16 +181,25 @@ class TestFirewall(NetworkCase):
...
-        b.click("#add-services-dialog .modal-footer .btn-primary")
-        b.wait_not_present("#cockpit_modal_dialog")
-        b.wait_present("tr[data-row-id='pop3']")
+        addService('pop3')
+        # firewalld > 0.7.0 supports including services into another service
+        if m.image in ['rhel-8-1']:
+            addService('freeipa-4')
+            b.click("tr[data-row-id='freeipa-4']")
+            b.click("tbody.open a:contains(kerberos)")
+            b.wait_present("tbody.open .listing-ct-body:contains(Included service)")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Please do an "else:" here, so that we have a reminder (failing test) once firewalld 0.7 trickles into distros. For example, [debian-testing](https://packages.debian.org/bullseye/firewalld) already has 0.7, so please add it here.

Comment (Gundersanne) [reply test/verify/check-networking-firewall]: I think our debian-testing is still buster?

**Comment by Gundersanne** [comment]
@garrett Would the updated photo's design work better?

**Comment by garrett** [comment]
What if the we ditched the tabs entirely and had the services listed under the details text?

Something like this:

---

> FreeIPA is an LDAP and Kerberos domain controller for Linux systems. Enable this option if you plan to provide a FreeIPA Domain Controller using the LDAP protocol. You can also enable the 'freeipa-ldaps' service if you want to provide the LDAPS protocol. Enable the 'dns' service if this FreeIPA server provides DNS services and 'freeipa-replication' service if this FreeIPA server is part of a multi-master replication setup.
>
> ### Included Services
>
> - **WWW (HTTP)**, port 80: HTTP is the protocol used to serve Web pages. If you plan to make your Web server publicly available, enable this option. This option is not required for viewing pages locally or developing Web pages. 
> - **Secure WWW (HTTPS)**, port 443: HTTPS is a modified HTTP used to serve Web pages when security is important. Examples are sites that require logins like stores or web mail. This option is not required for viewing pages locally or developing Web pages. You need the httpd package installed for this option to be useful. 
> - **Kerberos**, port 88: Kerberos network authentication protocol server Kpasswd: Kerberos password (Kpasswd) server 
> - **LDAP**, port 389: Lightweight Directory Access Protocol (LDAP) server 
> - **LDAPS**, port 636: Lightweight Directory Access Protocol (LDAP) over Secure Sockets Layer (SSL) server

---

It's just a heading (probably an h4 or h5 — whatever is 1 more than the previous heading), which would be styled to be no larger than the row's text (so we'll probably need to do `font-size: inherit` on it here), then a `<li>` with `<strong>` for each item's title.

I don't think it would be too tall to do it this way. As there's nothing but informative text, I'd like it if we could keep it all together in the details.

If we eventually add more features, we could then we could reconsider tabs. But I'm not sure what features that would be. For example: Editing is about which zone the service is in really... or if it's a custom rule, then the ports selected too. (In other words: editing would be in a dialog, not inline in an expanded row.) So if not editing, then what? This is why I don't think we need tabs here, either in the present or the future.

---

**TL;DR**: Bulleted list with a heading. Names of services would be in a `<strong>` tag. No tabs.

**Comment by marusak** [comment]
Looks great now! :) Both debian stable and testing failed in `testFirewallPage` exactly in the same way which seems suspicious. I've retried it, lets see

File: pkg/lib/cockpit-components-listing.jsx [hunk 3]
```diff
@@ -49,6 +49,8 @@ import './listing.less';
  * addCheckbox optional: if set a checkbox will appear in the start of the row and the selectChanged
  *                       callback can be used to track row's checked status. Note that rows with checkboxes can't
  *                       be selected outside of the checkbox.
+ * simpleBody optional: if set the expansion will just contain this simple body without tabs,
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing.jsx]: This should also be in `propTypes`

File: pkg/lib/cockpit-components-listing.jsx [hunk 4]
```diff
@@ -253,21 +255,30 @@ export class ListingRow extends React.Component {
                 );
             }
 
+            let simpleBody, heading;
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing.jsx]: Will `let simpleBody` set `simpleBody` to null/undefined? (I am asking for my own information :) )

File: pkg/lib/cockpit-components-listing.jsx [hunk 5]
```diff
@@ -253,21 +255,30 @@ export class ListingRow extends React.Component {
...
+            if ('simpleBody' in this.props) {
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing.jsx]: I think we usually do `if (this.props.simpleBody)`, not that it matters much :)

File: pkg/lib/cockpit-components-listing.jsx [hunk 6]
```diff
@@ -253,21 +255,30 @@ export class ListingRow extends React.Component {
...
-                                <div className="listing-ct-actions">
-                                    {listingDetail}
-                                    {this.props.listingActions}
-                                </div>
-                                <ul className="nav nav-tabs nav-tabs-pf">
-                                    {links}
-                                </ul>
-                            </div>
-                            {tabs}
+                            { tabs && heading }
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing.jsx]: this `tabs` just tests if tabs is defined and if is, then it shows heading? Wouldn't it make sense to just show heading when defined? (`{ heading }`)
Or wouldn something like `simpleBody || (heading && tabs)` work? (reading like "set simpleBody or heading followed by tabs")

**Comment by marusak** [comment]
I am afraid that tests are broken on debian. I've seen 4 results and all 4 failed the same way on [testFirewallPage](https://209.132.184.41:8493/logs/pull-12385-20190729-091818-c24449bc-cockpit-project-cockpit-debian-stable/log.html#238). Unfortunately needs work

File: pkg/lib/cockpit-components-listing.jsx (hunk 4 above)
Comment (Gundersanne) [reply pkg/lib/cockpit-components-listing.jsx]: Should be `undefined` yup

File: pkg/lib/cockpit-components-listing.jsx (hunk 5 above)
Comment (Gundersanne) [reply pkg/lib/cockpit-components-listing.jsx]: It's more because the exact same formulation is used for listingDetail just above it.

**Comment by andreasn** [comment]
> Is there a service that pulls in more services than FreeIPA currently?

Satellite does:
Screenshot from 2019-07-29 12-06-10

**Comment by garrett** [comment]
I'm still unable to try this out locally, as firewalld is too old in Fedora 30.

It was released into rawhide last week, but even rebuilding the package for F30 doesn't work here, as overlaying it on Silverblue won't work, as it's a base package and rpm-ostree forbids updating it manually.

I'm currently trying to get rawhide working enough on a test laptop to check out this branch.

**Comment by garrett** [comment]
I've adjusted the spacing of the bundled services in the services list and adjusted the listing in general.

Here's a screenshot:
Screenshot_2019-07-30 localhost localdomain

File: pkg/networkmanager/firewall.jsx [hunk 7]
```diff
@@ -90,14 +97,22 @@ function ServiceRow(props) {
...
+            if (service && service.description)
+                return <li key={service.id}><strong>{service.name}</strong>: {service.description}</li>;
+        })} </ul>);
+    }
 
     return <ListingRow key={props.service.id}
                        rowId={props.service.id}
                        columns={columns}
-                       tabRenderers={tabs} />;
+                       simpleBody={simpleBody.length > 0 ? simpleBody : null} />;
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: This means simpleBody is an array of JSX objects. With that they need `key=` properties, otherwise you get React warnings. They may be hidden by the fact that we only enable them and make them fatal on fedora-30 test runs, and this only runs on rhel-8-1 right now.

But instead of key= it may be better to restructure this and avoid the array altogether, split the simpleBody contents in two halves (set conditionally), and render them here in a `<React.Fragment>`.

**Comment by garrett** [comment]
@martinpitt: `lib/listing: Remove font px hardcoding` should probably be separate, as it affects all of Cockpit and is not (necessarily) related to this PR (but does affect it). But other commits should be merged.

**Comment by martinpitt** [comment]
@garrett: Indeed, my bad. (← @Gundersanne )

File: pkg/networkmanager/firewall.jsx [hunk 8]
```diff
@@ -90,14 +97,26 @@ function ServiceRow(props) {
...
+        includes = <React.Fragment>
+            <h5>Included Services</h5>
+            <ul>{props.service.includes.map(s => {
+                let service = firewall.services[s];
+                if (service && service.description)
+                    return <li key={service.id}><strong>{service.name}</strong>: {service.description}</li>;
+            })} </ul></React.Fragment>;
+    }
+    if (description || includes)
+        simpleBody = <React.Fragment>{description}{includes}</React.Fragment>;
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Would be a bit weird to have neither, but visually this shouldn't actually be different to what we have now.

**Comment by martinpitt** [comment]
Too many flakes, retrying a bit. Please self-land when all firewall tests pass.

File: pkg/networkmanager/firewall.jsx (hunk 8 above)
Comment (Gundersanne) [reply pkg/networkmanager/firewall.jsx]: When neither are present the row is simply not expandable. This is the case in custom ports.

**Comment by martinpitt** [comment]
Still [two](https://209.132.184.41:8493/logs/pull-12385-20190802-070101-49d3e7c7-cockpit-project-cockpit-debian-stable/log.html#238) [firewall](https://209.132.184.41:8493/logs/pull-12385-20190802-070101-49d3e7c7-cockpit-project-cockpit-rhel-8-1/log.html#235) failures, retrying more.

File: pkg/networkmanager/firewall-client.js [hunk 9]
```diff
@@ -167,21 +178,39 @@ function fetchServiceInfos(services) {
...
                     };
 
                     firewall.services[service] = info;
+                    return firewalld_dbus.call('/org/fedoraproject/FirewallD1/config',
+                                               'org.fedoraproject.FirewallD1.config',
+                                               'getServiceByName', [service]);
+                })
+                .then(path => firewalld_dbus.call(path[0],
+                                                  'org.fedoraproject.FirewallD1.config.service',
+                                                  'getSettings2', []))
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: Follow-up: Start with this call, and if it fails, fall back to the old one and set includes to empty. Saves one call on new firewalld, and on older versions there's no change (one succeeding and one failing call).

Comment (Gundersanne) [reply pkg/networkmanager/firewall-client.js]: Right, that would make a lot more sense. I'll do it as a followup, this one has been in the pipeline too long :p

**Comment by martinpitt** [comment]
@Gundersanne : can you please send a PR against the rhel-8.1 branch with this backported?

**Comment by Gundersanne** [comment]
@martinpitt  sure, will be my first backport :D

**Comment by martinpitt** [comment]
@Gundersanne: Sorry, can you please do the screenshot slightly wider (to avoid the label wrap) and have some more context? Right now it's hard to see from where this actually happens.

**Comment by Gundersanne** [comment]
@martinpitt 

image

image
